dm/syncer/data_validator.go

```diff
@@ -447,119 +611,143 @@ func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *re
 		return nil
 	}
 
-	targetTable := v.syncer.route(sourceTable)
-	// todo: syncer will change schema in schemaTracker, will there be data race?
-	// todo: what if table is dropped while validator falls behind?
-	tableInfo, err := v.syncer.schemaTracker.GetTableInfo(sourceTable)
+	table, err := v.genValidateTableInfo(sourceTable)
 	if err != nil {
 		if schema.IsTableNotExists(err) {
 			// not a table need to sync
 			return nil
+		} else if err == errNoPrimaryKey {
+			v.tableStatus[fullTableName] = &tableValidateStatus{
+				source:  *sourceTable,
+				target:  *table.Target,
+				stage:   pb.Stage_Stopped,
+				message: tableWithoutPrimaryKeyMsg,
+			}
+			return nil
 		}
 		return terror.Annotate(err, "failed to get table info")
 	}
 
-	if len(tableInfo.Columns) < int(ev.ColumnCount) {
-		v.unsupportedTable[fullTableName] = "binlog has more columns than current table"
-		return nil
-	}
-
-	var primaryIdx *model.IndexInfo
-	for _, idx := range tableInfo.Indices {
-		if idx.Primary {
-			primaryIdx = idx
+	if len(table.Info.Columns) < int(ev.ColumnCount) {
+		v.tableStatus[fullTableName] = &tableValidateStatus{
+			source:  *sourceTable,
+			target:  *table.Target,
+			stage:   pb.Stage_Stopped,
+			message: moreColumnInBinlogMsg,
```

Comment:
> I guess this is just a checker that has false negative?
since there's no ddl history now(not plan to add it in this version), we restrict that the table validated should not drop/ columns or switch order or ADD COLUMN AFTER.  not a false negative, we just cannot handle it now.